src/Acuminator/Acuminator.Analyzers/StaticAnalysis/DacFieldWithPXSelectorAndMismatchedType/DacFieldWithPXSelectorAndMismatchedTypeAnalyzer.cs

```diff
@@ -0,0 +1,152 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using Acuminator.Analyzers.StaticAnalysis.Dac;
+using Acuminator.Utilities.DiagnosticSuppression;
+using Acuminator.Utilities.Roslyn.Constants;
+using Acuminator.Utilities.Roslyn.PXFieldAttributes;
+using Acuminator.Utilities.Roslyn.Semantic;
+using Acuminator.Utilities.Roslyn.Semantic.Attribute;
+using Acuminator.Utilities.Roslyn.Semantic.Dac;
+using Acuminator.Utilities.Roslyn.Syntax;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Acuminator.Analyzers.StaticAnalysis.DacFieldWithPXSelectorAndMismatchedType
+{
+	public class DacFieldWithPXSelectorAndMismatchedTypeAnalyzer : DacAggregatedAnalyzerBase
+	{
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(
+				Descriptors.PX1078_TypesOfDacFieldAndReferencedFieldMismatch,
+				Descriptors.PX1078_TypesOfDacFieldAndReferencedFieldHaveDifferentSize);
+
+		public override bool ShouldAnalyze(PXContext pxContext, [NotNullWhen(true)] DacSemanticModel dac) => true;
+
+		public override void Analyze(SymbolAnalysisContext context, PXContext pxContext, DacSemanticModel dac)
+		{
+			context.CancellationToken.ThrowIfCancellationRequested();
+
+			var propertiesToCheck = dac.DeclaredDacFieldPropertiesWithAcumaticaAttributes
+				.Where(p => ContainsPxSelectorAttrbiute(pxContext, p));
+
+			foreach (var property in propertiesToCheck)
+			{
+				context.CancellationToken.ThrowIfCancellationRequested();
+
+				CheckProperty(context, property, dac, pxContext);
+			}
+		}
+
+		private static bool ContainsPxSelectorAttrbiute(PXContext pxContext, DacPropertyInfo property) =>
+			property.Attributes.Any(x => ContainsAttributeMatching(x, pxContext.AttributeTypes.PXSelectorAttribute.Type));
+
+		private static void CheckProperty(SymbolAnalysisContext context, DacPropertyInfo property, DacSemanticModel dac, PXContext pxContext)
+		{
+			var localType = property.Attributes.FirstOrDefault(
+				x => ContainsAttributeMatching(x, pxContext.AttributeTypes.PXDBFieldAttribute)
+				|| ContainsAttributeMatchingOneOf(x, pxContext.AttributeTypes.UnboundDataTypeAttributes));
+			var selector = property.Attributes.FirstOrDefault(x => ContainsAttributeMatching(x, pxContext.AttributeTypes.PXSelectorAttribute.Type));
+
+			if (selector is null || localType is null)
+				return;
+
+			var targetField = selector.AttributeData.ConstructorArguments.FirstOrDefault();
+
+			if (targetField is not { Kind: TypedConstantKind.Type, Value: ISymbol target }
+				|| target.ContainingSymbol is not INamedTypeSymbol parentType)
+				return;
+
+			var targetDac = DacSemanticModel.InferModel(pxContext, parentType, cancellation: context.CancellationToken);
+
+			if (targetDac?.PropertiesByNames.TryGetValue(target.Name, out var targetProperty) != true)
+				return;
+
```

Comment:
The point of diagnostic is to:
- check for size match (which can be checked for some numeric types as well)
- check that types match

Though it really may be simpler to check that CLR types match before going to deep into attribute check. A separate diagnostic descriptor will be needed, though

Reply:
I don't understand why check data attributes to verify if types are consistent instead of the direct check of property types. I also don't think that a third descriptor is needed. Imho, the check should go like this:
- check property types, report types inconsistency if they are not equal.
- for string property type look for data type attributes and check for length equality. Report length inequality if needed.
 
We check the consistency of property and DB column types, not consistency of data type attributes. I believe, the consistency of data type attributes is:
- redundant for the check of DAC field types compatibility
- not important for developers
- impossible to check in a general case because attribute can contain arbitrary code

Reply:
@acasj  I will try to explain these points. I will use the term C# type instead of CLR type but they mean the same thing.

**Redundancy**

If C# property types of DAC fields are not the same, then it doesn't matter, if data type attributes are the same or not. At least I'm not aware of a scenario where the field having a foreign key and the key field from the external DAC have the same data type attribute and different C# types. If there is such an example, please share it. 

If the CLR property types are the same, then in most cases DAC field types are compatible. The exception is when they have incompatible types of DB columns. That's why we are checking `string` DAC fields for string length. In theory, we could also check `decimal` and date time fields but I don't know any scenario where such fields are used to store a foreign key to another entity and a compatibility check is needed. In each of such cases there will be different data type attributes with different set of properties and constructor arguments. Each scenario will require a separate analysis. So, there is no need to check data type attributes.

It's important to note, that each data type attribute corresponds to only one C# type. It is probably possible to write data type attributes that works with different C# types, but fortunately there are no such attributes. 
 
If some developer declared on a DAC property a data type attribute incompatible with its C# type then Acuminator will issue an error because it already has diagnostic (https://github.com/Acumatica/Acuminator/blob/dev/docs/diagnostics/PX1021.md) that checks the compatibility of a property type with data type attributes declared on it.

So, I'll prove the equivalency starting from the opposite. Suppose that property types of a local and referenced DAC fields are not equal but the types deduced from their data type attributes are the same. Then there will be either a PX1021 error on one of the fields (and we can consider this case covered as an error), or a contradiction - each type defined by a data type attribute is compatible with its field's property type, each data type attribute corresponds to only one C# type, but C# types are not equal. Contradiction.

The opposite can be proved the same. If C# types are equal then types defined by data type attributes have to be equal.

Reply:
**Not important for developers**

This is just my opinion, but I believe that for developers in general it is not important what kind of data type attribute is declared on a referenced field. What matters to them is the consistency of fields' data types - both in DB and in memory. 

**Impossible to check**

There are many cases when it will be possible to check property types but impossible data type attributes, or such check will be incorrect. The most probable case - developer mistake, when developer forgot to define a data type attribute. 
Another scenario - it is possible in Acumatica to create your own data type attribute, for example for int fields. If such attribute does not derive from `PXDBInt`/`PXInt` the check for data type attributes will be incorrect.